source/core/types/ut_output_dbms_pipe.tpb

```diff
@@ -0,0 +1,76 @@
+create or replace type body ut_output_dbms_pipe as
+
+  constructor function ut_output_dbms_pipe(self in out nocopy ut_output_dbms_pipe) return self as result is
+  begin
+    self.output_type := $$plsql_unit;
+    self.output_id := self.generate_output_id;
+    return;
+  end;
+
+  overriding member procedure open(self in out nocopy ut_output_dbms_pipe) is
+    l_flag integer;
+  begin
+    --create an explicit private pipe. Explicitly created pipes need to be removed explicitly
+    --otherwise they stay in memmory forever https://docs.oracle.com/cd/B19306_01/appdev.102/b14258/d_pipe.htm#CHDEICJI
+    --to check if there are any non-purged pipes execute:  select * from v$db_pipes where pipe_size > 0 order by name desc;
+    l_flag := dbms_pipe.create_pipe(self.output_id);
+  end;
+
+  overriding member procedure send(self in out nocopy ut_output_dbms_pipe, a_text clob) is
+    c_size_limit_chars constant integer := ut_output_pipe_helper.gc_size_limit_chars;
+    l_text_part        ut_output_pipe_helper.t_pipe_item;
+    i                  integer := 0;
+  begin
+    --split test into pieces of a size valid for pipe and send to pipe
+    while i < length(a_text) loop
+      l_text_part := substr( a_text, i + 1, c_size_limit_chars );
+      ut_output_pipe_helper.send( self.output_id, l_text_part);
+      i := i + c_size_limit_chars;
+    end loop;
+    --SEND is closed by a EOM message
+    ut_output_pipe_helper.send( self.output_id, ut_output_pipe_helper.gc_output_eom );
```

Comment:
we could sent a NULL value with ROWID data type and a NULL value with a RAW data type.
Since pipe is used only to transfer text, types different than text would uniquely identify control communication.